[no-throw-literal] assumes intersection types are not errors

> Make sure you read through our FAQ before posting.
> https://github.com/typescript-eslint/typescript-eslint/blob/issue-template-update/docs/getting-started/linting/FAQ.md

_Note:_ ↑ Link is broken, I read [this](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md).

**Repro**


_.eslintrc_
```javascript
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: { project: './tsconfig.json' },
  plugins: [ '@typescript-eslint' ],
}
```

_test.ts_
```TS
/*eslint @typescript-eslint/no-throw-literal: "error"*/
function foo () {
  throw Object.assign(new Error('message'), { foo: 'bar' })
}
```

**Expected Result**

No error, as the union type of `Error | { foo: 'bar' }` does implement "Error"

**Actual Result**

```
  3:9  error  Expected an error object to be thrown  @typescript-eslint/no-throw-literal
```

**Additional Info**

This is somewhat related, but in typescript `Foo & Bar` specifies the api but doesn't specify how `instanceof` is supposed to behave. `Foo & Error` could be an instance of `Foo` or `Error` or something completely different. 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.26.0` |
| `@typescript-eslint/parser`        | `2.26.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.11.11` |
| `npm`                              | `6.14.1` |

--- a/packages/eslint-plugin/src/rules/no-throw-literal.ts
+++ b/packages/eslint-plugin/src/rules/no-throw-literal.ts
@@ -28,6 +28,13 @@ export default util.createRule({
     const checker = program.getTypeChecker();
 
     function isErrorLike(type: ts.Type): boolean {
+      if (type.isIntersection()) {
+        return type.types.some(isErrorLike);
+      }
+      if (type.isUnion()) {
+        return type.types.every(isErrorLike);
+      }
+
       const symbol = type.getSymbol();
       if (!symbol) {
         return false;

--- a/packages/eslint-plugin/tests/rules/no-throw-literal.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-throw-literal.test.ts
@@ -96,6 +96,19 @@ throw new CustomError();
 class CustomError<T extends object> extends Error {}
 throw new CustomError();
     `,
+    `
+function foo() {
+  throw Object.assign(new Error('message'), { foo: 'bar' });
+}
+    `,
+    {
+      code: `
+const foo: Error | SyntaxError = bar();
+function bar() {
+  throw foo;
+}
+      `,
+    },
   ],
   invalid: [
     {
@@ -349,6 +362,31 @@ function foo<T>(fn: () => Promise<T>) {
           messageId: 'object',
           line: 5,
           column: 9,
+        },
+      ],
+    },
+    {
+      code: `
+function foo() {
+  throw Object.assign({ foo: 'foo' }, { bar: 'bar' });
+}
+      `,
+      errors: [
+        {
+          messageId: 'object',
+        },
+      ],
+    },
+    {
+      code: `
+const foo: Error | { bar: string } = bar();
+function bar() {
+  throw foo;
+}
+      `,
+      errors: [
+        {
+          messageId: 'object',
         },
       ],
     },

